### DISCUSSION TITLE
Looping questions

### DISCUSSION THREAD

**Discussion by souravlal** [marcel-licence/esp32_midi_sampler#41] [Q&A]
The ESP32 audio kit outputs humming sound while playing notes. I have connected an esp8266 to send MIDI data to ESP32 audio kit. IO18 is used to receive MIDI data, ground is connected to ground. I have tried different amplifiers and power supplies, ground connections are ok , I don't think is a ground humming also don't think is a hardware problem. The humming sound appears only when the kit starts to play sound, when playing no tones there is no problem.
The tone also doesn't sound like a bottle sound its more like a drum sound, also the sound output is very low, even though I put the amplifier to maximum volume. I also tried the speaker output of the kit both channels are not working.

I used this code for esp8266:- https://github.com/rjobidon/esp8266-midi-out

**Comment by Slider2732** [comment]
Quick answer - have you tried moving the Audio Kit jack plug in and out slightly?
My set up has the 3.5mm plug halfway out, otherwise I was getting hum. Tried headphones and out to my audio recorder, both plugs needed to be halfway out. 
I think it depends on whether you have a 3 ring plug or 2 ring plug.

**Reply by souravlal** [reply]
I am using esp8266 to send midi data to ESP32 Audio-kit with your code.
Using this code for esp8266 :-https://github.com/rjobidon/esp8266-midi-out

**Reply by marcel-licence** [reply]
You could try to use a USB power bank to sort out that humming is introduced there. I am confused that there is no humming when you play no sounds, I've never seen that effect before. Do you have also humming when connecting headphones?

**Reply by souravlal** [reply]
Same humming sound with headphones.
I have tried different power supplies.

**Reply by marcel-licence** [reply]
Hm I have no idea. Do you have an audio sample to listen to?

**Reply by souravlal** [reply]
I have tried tweaking the position of the audio jack but no change.
This is how it sounds.👇

https://user-images.githubusercontent.com/53053463/127770445-98a8460c-a792-40e5-95c5-98a9bb8d4b76.mp4

**Reply by souravlal** [reply]
I have downgraded my esp32 cores to 1.0.4 from 1.0.6, but after that when I upload the code error messages occurs. Error messages mentioned below:- 



Arduino: 1.8.12 (Windows 8.1), Board: "ESP32 Dev Module, Enabled, Default 4MB with spiffs (1.2MB APP/1.5MB SPIFFS), 240MHz (WiFi/BT), QIO, 80MHz, 4MB (32Mb), 921600, None"

C:\Users\TOSHIBA\Documents\Arduino\libraries\Adafruit_NeoPixel\esp.c: In function 'espShow':

C:\Users\TOSHIBA\Documents\Arduino\libraries\Adafruit_NeoPixel\esp.c:100:43: error: missing binary operator before token "("

 #if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4, 0, 0)

                                           ^

C:\Users\TOSHIBA\Documents\Arduino\libraries\Adafruit_NeoPixel\esp.c:128:43: error: missing binary operator before token "("

 #if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4, 0, 0)

                                           ^

Multiple libraries were found for "WiFi.h"
 Used: C:\Users\TOSHIBA\AppData\Local\Arduino15\packages\esp32\hardware\esp32\1.0.4\libraries\WiFi
 Not used: C:\Program Files (x86)\Arduino\libraries\WiFi
exit status 1
Error compiling for board ESP32 Dev Module.

This report would have more information with
"Show verbose output during compilation"
option enabled in File -> Preferences.

**Reply by souravlal** [reply]
Is there any way I can play the sample from psram without midi controller for testing?
I have aslo replaced the 'bottle.wav' with a file I downloaded from internet(1second, 16 bit 44.1Khz, stereo) but that time no sound were made.

**Reply by marcel-licence** [reply]
`    /* use this to easily test the output */
#if 0
    Sampler_NoteOn(0, 64, 1);
#endif`
Ensure the sound is saved as mono and does not include additional tags (loading may fail). I would recommend open this with audcity, ensure that you have only one mono track and export this again.

**Reply by marcel-licence** [reply]
Thanks for the audio. I cannot hear the humming. It does play the sound twice. One is played using the loop continously (the loop sounds a bid odd). In addition to that I hear another sample playing in repeat. Downgrade shouldn't be required. I am also working with 1.0.6.
You should add some debug output when the NoteOn function is called to get an idea if there a multiple events.

**Reply by souravlal** [reply]
@marcel-licence I have disconnected my MIDI controller and  used the code you have mentioned

/* use this to easily test the output */
#if 1
Sampler_NoteOn(0, 64, 1);
#endif

Simply set it to '1' and here is the output :-👇


https://user-images.githubusercontent.com/53053463/127777922-98d9f411-242f-45ef-b313-5d9c6c3d36e6.mp4

This hum sound goes on and on until I power of the ESP32 Audio-kit.

**Reply by marcel-licence** [reply]
This is no hum. That is the voice as expected. It has a loop in it. So without NoteOff it will not turn off.

**Reply by marcel-licence** [reply]
https://youtu.be/7uSobNW7_A4?t=400 it is the bottle sound but 1-2 octaves lower

**Reply by souravlal** [reply]
@marcel-licence I have replaced the 'pet_bottle.wav' sample with my own sample, sampled at 44.1Khz, mono, 2sec and I exported it as  signed 16-bit pcm using audacity but cannot hear any sound.
Is there anything I need to do with the '.bin' file located at the same directory?
Do I need to change the file name in the code.

**Reply by marcel-licence** [reply]
You should delete the bin file because there is stored a configuration which was for the original sound. Without configuration a sample will be played as one shot from beginning to end.

**Reply by souravlal** [reply]
I have used simple delay of 2sec to send a note off and successfully played the 'pet_bottle.wav' thanks to your help @marcel-licence.
But when I used a higher note number like 100 the pitch doesn't hear like it is increased. Also the .wav audio sounds different while I played it on my computer using a media player.

**Reply by marcel-licence** [reply]
Use `Sampler_NoteOn(1, 64, 1);` <- should send on CH1 to allow pitching. Ch0 is used to play different samples at their original pitch.

**Reply by souravlal** [reply]
👍.
Can anyone suggest me a good MIDI in circuit using optocoupler. I have used 4N35 which has some delay issues.
6N138 not available in my region.

**Reply by souravlal** [reply]
@marcel-licence Does the pitch of the original sample wav file we upload to PSRAM matter to the pitch of the output audio we hear? Or it can be at any pitch(note)?

**Reply by marcel-licence** [reply]
There is somewhere a define which defines the note for the original pitch. Notes above will pitch up, notes below will pitch down. You can also modify the pitch via control change.

**Reply by marcel-licence** [reply]
For MIDI I am using the H11L1. You will find documentation when using google or other search engines

**Reply by souravlal** [reply]
@marcel-licence I have uploaded a sample to PSRAM and played it using note on and note off command with 3 sec delay in between at channel-1, I have played it with MIDI note number 60 and 100 but the one with high pitch(100) does not lasts for too long, it plays out very fast, no matter whatever delay I use or even if not use note off command it plays out immediately. Here is how both notes sounds like:-👇


https://user-images.githubusercontent.com/53053463/127902386-7446a666-baf6-409e-971d-6a327180479b.mp4


https://user-images.githubusercontent.com/53053463/127902394-d6ddc18d-3892-413e-b4e0-0be77cab5860.mp4

**Reply by marcel-licence** [reply]
Yeah thats correct. @ 44100 you have ~132300 samples when the sound is 3 seconds long. When playing with higher pitch also the speed will be increased. Changing pitch without speed requires a lot of computing effort. There are ways but not implemented in this project. You can now use the loop functionality. Use Sampler_LoopAll(1, 1) to loop the complete sample. Or you can set manually the loop start and end points. In that case it will be repeated until the NoteOff event appears.

**Reply by souravlal** [reply]
But in your  videos you are playing your MIDI controller with Audio-kit, it doesn't feels like there is a speed increase/decrease with different notes.

Sampler_LoopAll(1, 1)
Can I call this function anywhere in the code?
 What is (1,1) parameters in this function?

**Reply by marcel-licence** [reply]
```
void Sampler_LoopAll(uint8_t index, float value)
{
    if ((value > 0) && (lastActiveRec != NULL))
    {
        float sampleLen = (lastActiveRec->end - lastActiveRec->start);

        lastActiveRec->loop_start = 0;
        lastActiveRec->loop_end = sampleLen - 1;

        Status_ValueChangedFloat("len", lastActiveRec->end - lastActiveRec->start);
        Status_ValueChangedFloat("loop_start", lastActiveRec->loop_start);
        Status_ValueChangedFloat("loop_end", lastActiveRec->loop_end);
    }
}
```

**Reply by marcel-licence** [reply]
You should call NoteOn before to get this sample as active sample.

**Reply by souravlal** [reply]
Is setting loop time manually means this project will work like any normal keyboard @ any notes?
Don't you feel the difference in duration of each notes while playing and doesn't it effect your keyboard playing?

**Reply by souravlal** [reply]
@marcel-licence I have called "Sampler_LoopAll(1,1)" function after I called note on for MIDI note number 100 @ channel 2. But it doesn't seems to loop the sample l, in fact it triggers it multiple times. Here it how it sounds like:-👇


https://user-images.githubusercontent.com/53053463/128053865-a7167a24-cd2a-45f0-91ee-1a7e4249c3be.mp4

**Comment by Slider2732** [comment]
I've had great success looping using the pots via MIDI. It takes a while to get the sound without clicks etc, but works great. Sounds like yours is repeating the whole sample, not from and to somewhere in the middle.

A workaround, is to use the reverb and delay effects... the ear hears the time constants of the effects, after recognising the sound. As such, the note lengths are helped with a form of obfuscation.
For example, the Pet Bottle sample - your ears hear the bottle sound, but don't really have a reference for the length of that sound (unlike say a finger snap). With something staccato it would make little difference and is where the onboard envelope shaper can help too. 

I will say, sampling has always been that way to my own ears, shorter version higher up, longer down low, due to the amount of samples being played. A 1 second sample can last 30 seconds on the opposite end. Same since the 1980's! 
I'd think the sustain would be the parameter to adjust for constant length, but it's tough to get loop points correct 'on the fly'. Have to wonder though, if a loop point is found 'by ear' similarly to how Marcel has the code set up now, whether those same points could extend and truncate based on the keyboard notes being played. Sustain would be set with a specific amount of those loops per key, like a reference table. 

For your 6N138 troubles, have you tried car boot/yard sales in the local area? Old telephone gear, routers, PC power supplies or modems may well have them onboard.  A bit of desoldering salvage can sort many such supply troubles  :)

**Reply by souravlal** [reply]
Having said that with the default "Pett_bottle.wav" sample there is no duration issue for different notes. I have played it using MIDI note number 60 and 100, and here it how it sounds like:-

https://user-images.githubusercontent.com/53053463/128064443-832bb0df-30b0-4a42-9b4c-cb700fdd21c1.mp4

https://user-images.githubusercontent.com/53053463/128064753-714b2b72-f614-4eb6-bbe4-292f7ea5e2a7.mp4

But its a different story while I use my own samples, like I mentioned above.

**Reply by souravlal** [reply]
@Slider2732 Whatever samplers I have seen does not change speed of notes @ different pitch.
You said about looping, please help me, how can I change this? By doing this the duration(speed) problem can be fixed? Can it be done inside the code? Or is it done through switches attached using the help of display? Does setting loop needs to be done for each and every samples if I am playing from SD card?
MIDI circuit is not an issue, I can fix that out with some experiments, but what I struggle with is the software part(code).

**Reply by souravlal** [reply]
@marcel-licence  How do the default sample "pett_bottle.wav" plays without speed issues at all notes? Please let me know.

**Comment by souravlal** [comment]
Also a custom sample I have loaded to PSRAM stops playing without calling note off command.

**Comment by Slider2732** [comment]
I've not used pro modern samplers at all, experience is only with the Casio SK-5 or other vintage stuff. 
If time stretch is a common thing on pro gear then that's cool!
Though, my Zoom R8 recorder can stretch anything, to time or tempo. But, it takes 3 or 4 seconds to do so, for 1 sample... I expect it's all far too CPU intensive for real time on an ESP32, latency would be epic. 

No idea of how to implement the code, unfortunately.
Was just a time stretch idea  :)

**Reply by souravlal** [reply]
@Slider2732 But have you notice the case of the default sample "pett_bottle", it does play without any issues with duration of the notes at any notes. I have attached the audio above.
And should a note be continued playing until the noteoff command is send? This is not the case with my custom samples loaded to PSRAM with higher notes, what is your experience with SD card samples?

**Reply by marcel-licence** [reply]
By default, when loading just a wav file as sample it will play from begin to end as one shot when receiving a noteOn event. NoteOff will not have any effect if the release time is set to infinity (guess that is the default value). To have a stop you will need a release of 0.
The PetBottle sound has also a .bin file which has a configuration in it. There is a loop configured which is somewhere in the middle of the audio to repeat a short part.
In case a loop is configured (which can be done manually as explained above) after loading the wav file the sound will stay active until NoteOff will be called. Is the loop not in range of the sample (end point is after the end of the sample) the loop won't work.
As a first test you should setup a complete loop (there is also a function for it available). Keep in mind loop manipulation functions will only modify the last played note. So it will require a noteOn call (note Off can be called after) to manipulate a specific sample. After you can test the new setting calling NoteOn again.

**Reply by marcel-licence** [reply]
Regarding time stretching etc ... modern keyboards (for example Roland FA-06) play samples also faster when changing the pitch of the note. Sounds often consists of multiple samples (for example one per octave). This will reduce the unnatural effect. It would be possible to change speed without pitch and pitch without speed but as @Slider2732 mentioned that would be more CPU intensive. In addition to that you can get annoying effects with specific algorithms. So the sampler actually works like tape or a vinyl record. Pitch and speed belong together because internally only the samplerate will be manipulated to get this effect.

**Reply by souravlal** [reply]
@marcel-licence How can I set a loop? Which is the function for this? Can you give me an example? Will setting loop required to be done each time after loading a sample if I am loading from SD card?
How can I create a bin configuration file for a sample like you did for "pett_bottle"? Also if I am using SD card to store and load samples is it possible to create bin files for all samples in SD card?

**Reply by souravlal** [reply]
Can anyone help me please with the above?

**Reply by souravlal** [reply]
@marcel-licence I am loading sample from PSRAM and I have to change the sample name in the code each time I load my custom sample, is there any way to avoid that and load whatever sample I upload to PSRAM automatically  without changing the sample name in the code?

**Reply by marcel-licence** [reply]
The bin file will be written when saving a sample via 'Sampler_SavePatch'. It will store the actual configuration (loop etc.). 
Automatic loading would be possible. You should learn how to read a directory, extract the filenames. Finally you can use the list to execute the loading process for each file.

**Reply by souravlal** [reply]
@marcel-licence By saying "saving a sample via 'Sampler_SavePatch' ", you mean recording sound via mic or Audio input, Right?
I have used the 'Sampler_LoopAll(1, 1)' function but the sample seems to repeatedly trigger multiple times continuously.
How can I set loop manually. If yes, is it needed to do every time I load a sample from SD card?

**Reply by marcel-licence** [reply]
Looping a complete sample for beginning to end sounds like retriggering.
If you set loop points you can save the sample and a bin file will be created with the same name containing the configuration. This will be load if existing together with the new wav file.

**Reply by souravlal** [reply]
@marcel-licence How can I set a loop?
And how can I save a sample?where will it be saved?

**Reply by marcel-licence** [reply]
I am preparing a video soon 1-2 weeks if I can find enough time =) I am sure that will help you because I've found a way to integrate sf2 soundfonts using a little workaround

**Reply by souravlal** [reply]
Oh, looking forward to that. Good luck.

**Reply by marcel-licence** [reply]
This video: https://youtu.be/Accf54fJn6U may answer the questions regarding the setup of samples

**Reply by souravlal** [reply]
@marcel-licence Is it possible to set loop for whatever samples I want to put in to the sd card, using this technique?

**Reply by marcel-licence** [reply]
Sure this method is not limited to this example. You have only the limitation that the sampler project can only use one sample per instrument. The soundfont in general supports multiple samples per instrument for different pitches and velocity values. That is not supported by the sampler project. Another limitation is the 4MB PSRAM. The wav file shall be smaller than 4MB.
